### ISSUE TITLE
_handleUpload doesn't handle errors

### ISSUE THREAD

**Issue by SSSachith** [VeliovGroup/Meteor-Files#623]
Hi
I am using FilesCollection with a compound index that throws errors for duplicate names. But the error doesn't get propagate back to the user. Upon inspection with my colleague we noticed that _handleUpload swollows the error and does not pass it to the callback.
[link](https://github.com/VeliovGroup/Meteor-Files/blob/master/server.js#L507)

**Comment by dr-dimitru** [comment]
Hello @SSSachith ,

Please update this thread following our [issue template](https://github.com/VeliovGroup/Meteor-Files/blob/master/.github/ISSUE_TEMPLATE):

  - Version of `Meteor-Files` you're experiencing this issue
  - Version of `Meteor` you're experiencing this issue
  - Where this issue appears? OS (Mac/Win/Linux)? Browser name and its version?
  - Post *Server* logs with enabled `debug` option, you can enable "debug" mode in [*Constructor*](https://github.com/VeliovGroup/Meteor-Files/wiki/Constructor)

**Comment by SSSachith** [comment]
Thank you for the reply and sorry about not giving enough info. First time using github for this.
Meteor files version:
Meteor version: 1.7.0.3
Appears everywhere, tested in Window 10, chrome browser 66.0.3359.181 (Official Build) (64-bit) (cohort: Stable)

Basically we have a compound index that should cause insertion to fail if a file with the same name is inserted. The error is thrown in server side as it should. But it doesn't propagate back to the client. If you look at the _handleUpload in the link I have posted you can see the error needs to be passed into the call back.

**Comment by dr-dimitru** [comment]
@SSSachith thank you for update.
To fix your issue, first we need to replicate it, but still to move on we need:
 - *Server* logs with enabled `debug` option, you can enable "debug" mode in [*Constructor*](https://github.com/VeliovGroup/Meteor-Files/wiki/Constructor)
 - Index definition

**Comment by SSSachith** [comment]
Sure

**Server Log**
I20180625-11:05:46.180(10)? [FilesCollection] [Upload] [finish(ing)Upload] -> C:\GIT\dg-cms\MEDIA\kFBhewSrWtrQA9oYb.jpg
I20180625-11:05:46.374(10)? [FilesCollection] [Upload] [_finishUpload] [insert] Error: { WriteError({"code":11000,"index":0,"errmsg":"E11000 duplicate key error collection: meteor.Media index: meta.parentId_1_owner_1_name_1 dup key: { : \"QZbpSKoYifRQAYvAT\", : null, : \"slide2.jpg\" }","op":{"size":1922894,"type":"image/jpeg","name":"slide2.jpg","meta":{"name":"slide2.jpg","tags":[],"createdBy":"Cm3W5qtRaaq4Hhsby","isProcessedFile":false,"parentId":"QZbpSKoYifRQAYvAT"},"ext":"jpg","extension":"jpg","extensionWithDot":".jpg","mime":"image/jpeg","mime-type":"image/jpeg","_id":"kFBhewSrWtrQA9oYb","userId":"Cm3W5qtRaaq4Hhsby","path":"C:\\GIT\\dg-cms\\MEDIA\\kFBhewSrWtrQA9oYb.jpg","versions":{"original":{"path":"C:\\GIT\\dg-cms\\MEDIA\\kFBhewSrWtrQA9oYb.jpg","size":1922894,"type":"image/jpeg","extension":"jpg"}},"_downloadRoute":"/cdn/storage","_collectionName":"Media","isVideo":false,"isAudio":false,"isImage":true,"isText":false,"isJSON":false,"isPDF":false,"_storagePath":"C:\\GIT\\dg-cms\\MEDIA","public":false}})
I20180625-11:05:46.382(10)?     at resultHandler (C:\Users\sachi\AppData\Local\.meteor\packages\npm-mongo\3.0.7\npm\node_modules\mongodb\lib\bulk\ordered.js:459:11)
I20180625-11:05:46.384(10)?     at C:\Users\sachi\AppData\Local\.meteor\packages\npm-mongo\3.0.7\npm\node_modules\mongodb-core\lib\connection\pool.js:544:18
I20180625-11:05:46.385(10)?     at _combinedTickCallback (internal/process/next_tick.js:131:7)
I20180625-11:05:46.386(10)?     at process._tickDomainCallback (internal/process/next_tick.js:218:9)
I20180625-11:05:46.388(10)?   name: 'BulkWriteError',
I20180625-11:05:46.390(10)?   message: 'E11000 duplicate key error collection: meteor.Media index: meta.parentId_1_owner_1_name_1 dup key: { : "QZbpSKoYifRQAYvAT", : null, : "slide2.jpg" }',
I20180625-11:05:46.392(10)?   driver: true,
I20180625-11:05:46.393(10)?   code: 11000,
I20180625-11:05:46.395(10)?   index: 0,
I20180625-11:05:46.396(10)?   errmsg: 'E11000 duplicate key error collection: meteor.Media index: meta.parentId_1_owner_1_name_1 dup key: { : "QZbpSKoYifRQAYvAT", : null, : "slide2.jpg" }',
I20180625-11:05:46.397(10)?   getOperation: [Function],
I20180625-11:05:46.398(10)?   toJSON: [Function],
I20180625-11:05:46.400(10)?   toString: [Function],
I20180625-11:05:46.401(10)?   result:
I20180625-11:05:46.403(10)?    BulkWriteResult {
I20180625-11:05:46.404(10)?      ok: [Getter],
I20180625-11:05:46.406(10)?      nInserted: [Getter],
I20180625-11:05:46.407(10)?      nUpserted: [Getter],
I20180625-11:05:46.408(10)?      nMatched: [Getter],
I20180625-11:05:46.411(10)?      nModified: [Getter],
I20180625-11:05:46.413(10)?      nRemoved: [Getter],
I20180625-11:05:46.414(10)?      getInsertedIds: [Function],
I20180625-11:05:46.415(10)?      getUpsertedIds: [Function],
I20180625-11:05:46.417(10)?      getUpsertedIdAt: [Function],
I20180625-11:05:46.418(10)?      getRawResponse: [Function],
I20180625-11:05:46.419(10)?      hasWriteErrors: [Function],
I20180625-11:05:46.419(10)?      getWriteErrorCount: [Function],
I20180625-11:05:46.430(10)?      getWriteErrorAt: [Function],
I20180625-11:05:46.431(10)?      getWriteErrors: [Function],
I20180625-11:05:46.432(10)?      getLastOp: [Function],
I20180625-11:05:46.433(10)?      getWriteConcernError: [Function],
I20180625-11:05:46.433(10)?      toJSON: [Function],
I20180625-11:05:46.434(10)?      toString: [Function],
I20180625-11:05:46.435(10)?      isOk: [Function] } }


**Index Definition**
MediaCollection.collection
        .rawCollection()
        .createIndex(
            { 'meta.parentId': 1, owner: 1, name: 1 },
            { unique: true }
);

**MediaCollection Definition**
export const MediaCollection = new FilesCollection({
    storagePath,
    collectionName: 'Media',
    allowClientCode: false, 
    debug: true,
    onBeforeUpload(file) {
        try {
            mediaSchema
                .pick('meta', 'type')
                .validate({ meta: file.meta, type: file.type });
        } catch (err) {
            return err.toString();
        }
        return true;
    },
    onAfterUpload: function(fileRef) {
        if (fileRef.type === fileFormatEnum.MP4) {
            makeThumbFromVideo(fileRef);
        } else {
            makeThumbFromImage(fileRef);
            if (fileRef.meta.isProcessedFile) {
                resizeImage(fileRef);
            }
        }
    },
});

Thanks.

**Comment by coagmano** [comment]
Hi @dr-dimitru, to explain a bit further (I work with @SSSachith):

The insert that happens in `_finishUpload` is throwing an expected error, which is logged and passed to the callback here: [server.js#L921-L924](https://github.com/VeliovGroup/Meteor-Files/blob/master/server.js#L921-L924)

To find the callback which there error is passed to, we followed the call stack backwards:

`_finishUpload` is called by an event emitted in `_handleUpload`

When `opts.eof` is true `_handleUpload` is called from [server#L507](https://github.com/VeliovGroup/Meteor-Files/blob/master/server.js#L507) for HTTP transport and [server#L778](https://github.com/VeliovGroup/Meteor-Files/blob/master/server.js#L778) for DDP.

DDP works as expected as Line 778 will throw the error via `Meteor.wrapAsync`, which is caught and re-thrown, making it back to the client.

However for HTTP, the callback created at Line 507 doesn't accept any arguments and doesn't check if it was passed an error. This is where the error is swallowed and where we would like it to return the error to the client.

We believe the solution is something like this:
```javascript
this._handleUpload(result, opts, (error) => {
  if (error) {
    if (!httpResp.headersSent) {
      httpResp.writeHead(500);
    }
    if (!httpResp.finished) {
      httpResp.end(JSON.stringify(error));
    }
    return;
  }
  ...
```
But haven't checked what needs to be sent to the client so it picks up the error correctly or if this works for us yet

**Comment by dr-dimitru** [comment]
@coagmano @SSSachith thank you for the patience. Fix published in [`v1.10.0`](https://github.com/VeliovGroup/Meteor-Files/releases/tag/1.10.0). Let me know how it works on your end. Thank you a lot for contribution.
